soccer/src/rj_vision_filter/VisionFilter.cpp

```diff
@@ -1,76 +1,92 @@
+#include <rj_utils/logging.hpp>
 #include <rj_vision_filter/VisionFilter.hpp>
-#include <rj_vision_filter/util/VisionFilterConfig.hpp>
-
-VisionFilter::VisionFilter(Context* context) : context_{context} {
-    detection_frame_sub_ = std::make_unique<ros2_temp::DetectionFrameSub>();
-
-    threadEnd.store(false, std::memory_order::memory_order_seq_cst);
-
-    // Have to be careful so the entire initialization list
-    // is created before the thread starts
-    worker = std::thread(&VisionFilter::updateLoop, this);
+#include <rj_vision_filter/params.hpp>
+
+namespace vision_filter {
+DEFINE_FLOAT64(kVisionFilterParamModule, publish_hz, 120.0,
+               "The rate in Hz at which VisionFilter publishes ball and robot "
+               "observations.")
+
+VisionFilter::VisionFilter(const rclcpp::NodeOptions& options)
+    : rclcpp::Node{"vision_filter", options},
+      config_client_{this},
+      team_color_queue_{this, referee::topics::kTeamColorPub,
+                        rj_msgs::build<TeamColorMsg>().is_blue(true)},
+      param_provider_{this, kVisionFilterParamModule} {
+    // Create a timer that publishes the current state of the ball + robots.
+    publish_timer_callback_group_ = this->create_callback_group(
+        rclcpp::CallbackGroupType::MutuallyExclusive);
+    const std::chrono::duration<double> publish_timer_period(1 /
+                                                             PARAM_publish_hz);
+    publish_timer_ = create_wall_timer(
+        publish_timer_period, [this]() { PublishState(); },
+        publish_timer_callback_group_);
+
+    // Create a timer that calls predict on all of the Kalman filters.
+    predict_timer_callback_group_ = this->create_callback_group(
+        rclcpp::CallbackGroupType::MutuallyExclusive);
+    const std::chrono::duration<double> predict_timer_period(
+        PARAM_vision_loop_dt);
+    auto predict_callback = [this]() { PredictStates(); };
+    predict_timer_ = create_wall_timer(publish_timer_period, predict_callback,
+                                       predict_timer_callback_group_);
+
+    // Create a subscriber for the DetectionFrameMsg
+    update_callback_group_ = this->create_callback_group(
+        rclcpp::CallbackGroupType::MutuallyExclusive);
+    rclcpp::SubscriptionOptions detection_frame_sub_opts{};
+    detection_frame_sub_opts.callback_group = update_callback_group_;
+    constexpr int kQueueSize = 10;
+    const auto callback = [this](DetectionFrameMsg::UniquePtr msg) {
+        detection_frame_queue_.Push(std::move(msg));
+    };
+    detection_frame_sub_ = create_subscription<DetectionFrameMsg>(
+        vision_receiver::topics::kDetectionFramePub, rclcpp::QoS(kQueueSize),
+        callback, detection_frame_sub_opts);
+
+    // Create publishers.
+    world_state_pub_ =
+        create_publisher<WorldStateMsg>(topics::kWorldStatePub, 10);
+    last_updated_pub_ = create_publisher<TimeMsg>(topics::kLastUpdatedPub, 10);
```

Comment:
Can you explicitly make a QoS object here? I find it much more clear what the `10` parameter is when it's explicit in the type.

Reply:
Done.